guides/python-testing.md

```diff
@@ -0,0 +1,138 @@
+# Testing Python Packages
+
+The following guide is to layout a general guideline for testing Python
+packages. This document is not to be followed exactly word-by-word, but rather
+it lays out multiple things you should consider. Feel free to adopt and adapt
+as you deem necessary.
+
+## Do I need to test my Python package?
+
+Yes. Next question.
+
+## What should I use to run my tests?
+
+[pytest](https://docs.pytest.org/en/latest/) is recommended.
+There are many ``pytest`` plugins that you could utilize to get the most
+out of your tests. Example of several useful ones:
+
+* [pytest-doctestplus](https://github.com/astropy/pytest-doctestplus) for
+  advanced features for testing example code in documentation.
+* [pytest-openfiles](https://github.com/astropy/pytest-openfiles) for
+  detecting file handles that were inadvertently left open at the end of
+  unit tests.
+* [pytest-arraydiff](https://github.com/astrofrog/pytest-arraydiff) for
+  generation and comparison of data arrays produced during unit tests.
+* [pytest-socket](https://github.com/miketheman/pytest-socket) for handling
+  network calls.
+* [pytest-faulthandler](https://github.com/pytest-dev/pytest-faulthandler)
+  for getting a useful traceback out of segmentation fault.
+* [ci-watson](https://github.com/spacetelescope/ci_watson) for functionality
+  related to STScI Jenkins and Artifactory usage.
+
+## Where should I put my tests?
+
+See the templates in Python Packaging section.
+
+Template aside, there is also the question of whether you want to bundle the
+tests with your package or not. While bundling allows users to test their
+installations, it also increases the package size. For more information,
+see "Choosing a test layout / import rules" in
+https://docs.pytest.org/en/latest/goodpractices.html .
```

Comment:
There are maybe two separate decisions to be made here: 
1. Should the tests live in the source tree or outside of it entirely?
2. If the tests live outside of the source tree, should they still be included in the package distribution?

Reply:
Do you have a short text suggestion that would address these questions? I do not want this document to be a lengthy philosophical essay. Just short and sweet for people to get started with. And if they are interested to learn more, there is always the Internet.

Reply:
Feel free to edit this, Dan. You are a co-author. :wink:  :pray:  Maybe I missed something. Do we need to talk about `tox`?